contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -166,47 +331,126 @@ contract MixinMatchOrders is
-            matchedFillResults.right.takerAssetFilledAmount = rightTakerAssetAmountRemaining;
-            matchedFillResults.left.takerAssetFilledAmount = matchedFillResults.right.makerAssetFilledAmount;
-            // Round down to ensure the maker's exchange rate does not exceed the price specified by the order.
-            // We favor the maker when the exchange rate must be rounded.
-            matchedFillResults.left.makerAssetFilledAmount = _safeGetPartialAmountFloor(
-                leftOrder.makerAssetAmount,
-                leftOrder.takerAssetAmount,
-                matchedFillResults.left.takerAssetFilledAmount
-            );
+        // Maximally fill the orders and pay out profits to the matcher in one or both of the maker assets.
+        if (withMaximalFill) {
+            bool leftSpread;
+            bool rightSpread;
+
+            // Calculate the maximum fill results for the maker and taker assets. At least one of the orders will be fully filled.
+            //
+            // The maximum that the left maker can possibly buy is the amount that the right order can sell.
+            // The maximum that the right maker can possibly buy is the amount that the left order can sell.
+            //
+            // If the left order is fully filled, profit will be paid out in the left maker asset. If the right order is fully filled,
+            // the profit will be out in the right maker asset.
+            //
+            // There are three cases to consider:
+            // Case 1.
+            //   If the left maker can buy more than the right maker can sell, then only the right order is fully filled.
+            // Case 2.
+            //   If the right maker can buy more than the left maker can sell, then only the right order is fully filled.
+            // Case 3.
+            //   If the right maker can sell the max of what the left maker can buy and the left maker can sell the max of
+            //   what the right maker can buy, then both orders are fully filled.
+            if (leftTakerAssetAmountRemaining > rightMakerAssetAmountRemaining) {
+                // Case 1: Right order is fully filled with the profit paid in the left makerAsset
+                matchedFillResults.right.makerAssetFilledAmount = rightMakerAssetAmountRemaining;
+                matchedFillResults.right.takerAssetFilledAmount = rightTakerAssetAmountRemaining;
+                matchedFillResults.left.takerAssetFilledAmount = rightMakerAssetAmountRemaining;
+                // Round down to ensure the left maker's exchange rate does not exceed the price specified by the order.
+                // We favor the left maker when the exchange rate must be rounded and the profit is being paid in the
+                // left maker asset.
+                matchedFillResults.left.makerAssetFilledAmount = _safeGetPartialAmountFloor(
+                    leftMakerAssetAmountRemaining,
```

Comment:
Since we use the latter in the bottom else, I'm going to switch it for now. Since the latter is the price that the maker actually selected, I think it makes more sense to use that for approximations.